### ISSUE TITLE
need to navigate back three times to return back one page

### ISSUE THREAD

**Issue by metajack** [servo/servo#18564]
STR:

1. Visit https://metajack.im
2. Click the link to the LCA talk
3. Attempt to go back to original page. It currently takes me 3 clicks of the back button. The first click does nothing, the second removes the image preview for the embed, and the third finally returns to the home page.

**Comment by metajack** [comment]
cc @asajeffrey

**Comment by metajack** [comment]
This was found in servoshell, not sure if it's specific to that.

**Comment by asajeffrey** [comment]
Interesting. This might actually be doing the Spec Compatible Thing depending on how the youtube iframe is behaving: if it's navigating itself without enabling replacement, then it's meant to add a session history entry each time. Of course this isn't what the user (or content author) might be expecting :/

**Comment by metajack** [comment]
Firefox and Chrome both do the thing I expect (immediate navigation to home page), but they also support the video, so maybe this is just an issue in the fallback path. However, the next navigation after I get back always crashes Servo. See #18563

**Comment by asajeffrey** [comment]
Yes, this looks like a bug with how we're handling the session history, there should only be one entry created, since there's just the iframe's initial about:blank then the embedded video.

cc @cbrewster

**Comment by asajeffrey** [comment]
TMI:
```
$ RUST_LOG=constellation,script::dom::htmliframeelement ./mach run -d https://metajack.im
DEBUG:constellation::constellation: Adding new host entry metajack.im for top-level browsing context (0,2).
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: constellation got initiate navigate request message
DEBUG:constellation::constellation: constellation got set visibility change complete message
DEBUG:constellation::constellation: constellation got set final url message
DEBUG:constellation::constellation: Sending log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Received log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: constellation got activate document message
DEBUG:constellation::constellation: Document ready to activate (0,1)
DEBUG:constellation::constellation: Setting browsing context (0,2) to be pipeline (0,1).
WARN:constellation::browsingcontext: BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(0), index: BrowsingContextIndex(NonZero(2)) } iterated after closure.
DEBUG:constellation::constellation: Sending log entry Warn("BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(0), index: BrowsingContextIndex(NonZero(2)) } iterated after closure.").
DEBUG:constellation::constellation: Adding pipeline to new browsing context.
DEBUG:constellation::constellation: Creating new browsing context (0,2)
DEBUG:constellation::constellation: Setting activity of (0,1) to be FullyActive.
DEBUG:constellation::constellation: mozbrowser location change on a regular iframe (0,2)
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: Received log entry Warn("BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(0), index: BrowsingContextIndex(NonZero(2)) } iterated after closure.").
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
WARNING: cdn.mathjax.org has been retired. Check https://www.mathjax.org/cdn-shutting-down/ for migration tips.
DEBUG:constellation::constellation: constellation got head parsed message
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got joint session history length message from script
DEBUG:constellation::constellation: constellation got joint session history length message from script
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got load complete message
DEBUG:constellation::constellation: Pipeline (0,1) has no parent.
DEBUG:constellation::constellation: constellation got NodeStatus message
DEBUG:constellation::constellation: constellation got URL load message from script
DEBUG:constellation::constellation: Loading https://metajack.im/2017/01/18/servo-talk-at-lca-2017/ in pipeline (0,1).
DEBUG:constellation::constellation: Creating new pipeline (0,3) in browsing context (0,2).
DEBUG:constellation::constellation: constellation got initiate navigate request message
DEBUG:constellation::constellation: constellation got set visibility change complete message
DEBUG:constellation::constellation: constellation got set final url message
DEBUG:constellation::constellation: Sending log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Received log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: constellation got activate document message
DEBUG:constellation::constellation: Document ready to activate (0,3)
DEBUG:constellation::constellation: Setting browsing context (0,2) to be pipeline (0,3).
DEBUG:constellation::constellation: Adding pipeline to existing browsing context.
DEBUG:constellation::constellation: Setting activity of (0,1) to be Inactive.
DEBUG:constellation::constellation: Setting activity of (0,3) to be FullyActive.
DEBUG:constellation::constellation: mozbrowser location change on a regular iframe (0,2)
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
WARNING: cdn.mathjax.org has been retired. Check https://www.mathjax.org/cdn-shutting-down/ for migration tips.
DEBUG:constellation::constellation: constellation got head parsed message
DEBUG:script::dom::htmliframeelement: iframe bound to browsing context.
DEBUG:constellation::constellation: constellation got loaded `about:blank` in iframe message PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(NonZero(3)) } PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(NonZero(2)) }
DEBUG:constellation::constellation: constellation got set final url message
DEBUG:constellation::constellation: constellation got activate document message
DEBUG:constellation::constellation: Document ready to activate (1,2)
DEBUG:constellation::constellation: Sending log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Setting browsing context (1,1) to be pipeline (1,2).
WARN:constellation::browsingcontext: BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(1), index: BrowsingContextIndex(NonZero(1)) } iterated after closure.
DEBUG:constellation::constellation: Sending log entry Warn("BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(1), index: BrowsingContextIndex(NonZero(1)) } iterated after closure.").
DEBUG:constellation::constellation: Adding pipeline to new browsing context.
DEBUG:constellation::constellation: Creating new browsing context (1,1)
DEBUG:constellation::constellation: Setting activity of (1,2) to be FullyActive.
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: Received log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Received log entry Warn("BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(1), index: BrowsingContextIndex(NonZero(1)) } iterated after closure.").
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
DEBUG:constellation::constellation: constellation got head parsed message
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got iframe URL load message PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(NonZero(3)) } Some(PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(NonZero(2)) }) PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(NonZero(3)) }
DEBUG:constellation::constellation: Creating new pipeline (1,3) in browsing context (1,1).
DEBUG:constellation::constellation: Adding new host entry youtube.com for top-level browsing context (0,2).
DEBUG:constellation::constellation: constellation got initiate navigate request message
DEBUG:constellation::constellation: constellation got set visibility change complete message
DEBUG:constellation::constellation: constellation got set visibility change complete message
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got set final url message
DEBUG:constellation::constellation: Sending log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Received log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: constellation got activate document message
DEBUG:constellation::constellation: Document ready to activate (1,3)
DEBUG:constellation::constellation: Setting browsing context (1,1) to be pipeline (1,3).
DEBUG:constellation::constellation: Adding pipeline to existing browsing context.
DEBUG:constellation::constellation: Setting activity of (1,2) to be Inactive.
DEBUG:constellation::constellation: Setting activity of (1,3) to be FullyActive.
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
DEBUG:constellation::constellation: constellation got head parsed message
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got load complete message
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got NodeStatus message
DEBUG:constellation::constellation: Sending log entry Error("Error at :0:0 uncaught exception: wqa@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3996:169\nuqa@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3993:59\nB0@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3991:30\nm1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4071:150\nw1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4126:65\nz1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4138:118\n@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:7261:161\nGg/a.D<@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:379:223\nGg@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:380:10\nl.da@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:375:6\nPg/<@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:403:245\n").
DEBUG:constellation::constellation: Received log entry Error("Error at :0:0 uncaught exception: wqa@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3996:169\nuqa@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3993:59\nB0@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3991:30\nm1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4071:150\nw1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4126:65\nz1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4138:118\n@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:7261:161\nGg/a.D<@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:379:223\nGg@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:380:10\nl.da@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:375:6\nPg/<@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:403:245\n").
DEBUG:script::dom::htmliframeelement: iframe bound to browsing context.
DEBUG:constellation::constellation: constellation got loaded `about:blank` in iframe message PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(NonZero(3)) } PipelineId { namespace_id: PipelineNamespaceId(3), index: PipelineIndex(NonZero(3)) }
DEBUG:constellation::constellation: constellation got set final url message
DEBUG:constellation::constellation: constellation got activate document message
DEBUG:constellation::constellation: Document ready to activate (3,3)
DEBUG:constellation::constellation: Sending log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Setting browsing context (3,2) to be pipeline (3,3).
WARN:constellation::browsingcontext: BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(3), index: BrowsingContextIndex(NonZero(2)) } iterated after closure.
DEBUG:constellation::constellation: Sending log entry Warn("BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(3), index: BrowsingContextIndex(NonZero(2)) } iterated after closure.").
DEBUG:constellation::constellation: Adding pipeline to new browsing context.
DEBUG:constellation::constellation: Creating new browsing context (3,2)
DEBUG:constellation::constellation: Setting activity of (3,3) to be FullyActive.
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: Received log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Received log entry Warn("BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(3), index: BrowsingContextIndex(NonZero(2)) } iterated after closure.").
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
DEBUG:constellation::constellation: constellation got head parsed message
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got iframe URL load message PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(NonZero(3)) } Some(PipelineId { namespace_id: PipelineNamespaceId(3), index: PipelineIndex(NonZero(3)) }) PipelineId { namespace_id: PipelineNamespaceId(3), index: PipelineIndex(NonZero(4)) }
DEBUG:constellation::constellation: Creating new pipeline (3,4) in browsing context (3,2).
DEBUG:constellation::constellation: constellation got head parsed message
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got set final url message
DEBUG:constellation::constellation: Sending log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: constellation got activate document message
DEBUG:constellation::constellation: Document ready to activate (3,4)
DEBUG:constellation::constellation: Setting browsing context (3,2) to be pipeline (3,4).
DEBUG:constellation::constellation: Adding pipeline to existing browsing context.
DEBUG:constellation::constellation: Setting activity of (3,3) to be Inactive.
DEBUG:constellation::constellation: Setting activity of (3,4) to be FullyActive.
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: Received log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: constellation got head parsed message
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got set visibility change complete message
DEBUG:constellation::constellation: constellation got set visibility change complete message
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got load complete message
DEBUG:script::dom::htmliframeelement: iframe bound to browsing context.
DEBUG:constellation::constellation: constellation got loaded `about:blank` in iframe message PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(NonZero(3)) } PipelineId { namespace_id: PipelineNamespaceId(3), index: PipelineIndex(NonZero(6)) }
DEBUG:constellation::constellation: constellation got set final url message
DEBUG:constellation::constellation: constellation got activate document message
DEBUG:constellation::constellation: Document ready to activate (3,6)
DEBUG:constellation::constellation: Sending log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Setting browsing context (3,5) to be pipeline (3,6).
WARN:constellation::browsingcontext: BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(3), index: BrowsingContextIndex(NonZero(5)) } iterated after closure.
DEBUG:constellation::constellation: Sending log entry Warn("BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(3), index: BrowsingContextIndex(NonZero(5)) } iterated after closure.").
DEBUG:constellation::constellation: Adding pipeline to new browsing context.
DEBUG:constellation::constellation: Creating new browsing context (3,5)
DEBUG:constellation::constellation: Setting activity of (3,6) to be FullyActive.
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: Received log entry Warn("Resuming an already resumed timer.").
DEBUG:constellation::constellation: Received log entry Warn("BrowsingContext BrowsingContextId { namespace_id: PipelineNamespaceId(3), index: BrowsingContextIndex(NonZero(5)) } iterated after closure.").
DEBUG:constellation::constellation: constellation got head parsed message
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:script::dom::htmliframeelement: Unbinding frame (3,5).
DEBUG:constellation::constellation: constellation got remove iframe message
DEBUG:constellation::constellation: Closing browsing context (3,5).
DEBUG:constellation::constellation: Closing browsing context children (3,5).
DEBUG:constellation::constellation: Closing pipeline PipelineId { namespace_id: PipelineNamespaceId(3), index: PipelineIndex(NonZero(6)) }.
DEBUG:constellation::pipeline: pipeline PipelineId { namespace_id: PipelineNamespaceId(3), index: PipelineIndex(NonZero(6)) } exiting
DEBUG:constellation::constellation: Closed pipeline PipelineId { namespace_id: PipelineNamespaceId(3), index: PipelineIndex(NonZero(6)) }.
DEBUG:constellation::constellation: Closed browsing context children (3,5).
DEBUG:constellation::constellation: Closed browsing context BrowsingContextId { namespace_id: PipelineNamespaceId(3), index: BrowsingContextIndex(NonZero(5)) }.
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
DEBUG:script::dom::htmliframeelement: Discarding browsing context for pipeline (3,6)
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
DEBUG:constellation::constellation: constellation got set visibility change complete message
DEBUG:constellation::constellation: Pipeline PipelineId { namespace_id: PipelineNamespaceId(3), index: PipelineIndex(NonZero(6)) } exited.
DEBUG:constellation::constellation: constellation got load complete message
DEBUG:constellation::constellation: constellation got load complete message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got key event message
DEBUG:constellation::constellation: constellation got key event message
DEBUG:constellation::constellation: constellation got key event message
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got key event message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got traverse history message from compositor
DEBUG:constellation::constellation: Setting activity of (3,4) to be Inactive.
DEBUG:constellation::constellation: Setting activity of (3,3) to be FullyActive.
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got load complete message
DEBUG:constellation::constellation: Pipeline (0,3) has no parent.
DEBUG:constellation::constellation: Sending log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: Received log entry Warn("stop_parsing not implemented, full speed ahead!").
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got key event message
DEBUG:constellation::constellation: constellation got key event message
DEBUG:constellation::constellation: constellation got traverse history message from compositor
DEBUG:constellation::constellation: Setting activity of (1,3) to be Inactive.
DEBUG:constellation::constellation: Setting activity of (3,3) to be Active.
DEBUG:constellation::constellation: Setting activity of (1,2) to be FullyActive.
DEBUG:constellation::constellation: Sending frame tree for browsing context (0,2).
DEBUG:constellation::constellation: constellation got key event message
DEBUG:constellation::constellation: constellation got window resize message
DEBUG:constellation::constellation: handle_window_size_msg: 1024×740
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got iframe size message
DEBUG:constellation::constellation: constellation got a pending paint metric message
DEBUG:constellation::constellation: constellation got key event message
DEBUG:constellation::constellation: constellation got traverse history message from compositor
DEBUG:constellation::constellation: Setting activity of (0,3) to be Inactive.
DEBUG:constellation::constellation: Setting activity of (1,2) to be Active.
DEBUG:constellation::constellation: Setting activity of (0,1) to be FullyActive.
```

**Comment by asajeffrey** [comment]
Hmm I see three `DEBUG:script::dom::htmliframeelement: iframe bound to browsing context.` where I'd expect that there's just one. The second one is created right after:

```
DEBUG:constellation::constellation: Received log entry Error("Error at :0:0 uncaught exception: wqa@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3996:169\nuqa@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3993:59\nB0@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:3991:30\nm1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4071:150\nw1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4126:65\nz1@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:4138:118\n@https://www.youtube.com/yts/jsbin/player-vfliXTNRk/en_US/base.js:7261:161\nGg/a.D<@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:379:223\nGg@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:380:10\nl.da@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:375:6\nPg/<@https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js:403:245\n").
```

It could be that youtube's JS has a fallback path where it creates extra iframes, which is causing the excess session history entries?

**Comment by metajack** [comment]
@cpearce Is there a way to disable gecko's media stack so that YouTube embeds would behave more like current Servo to test the fallback path?

**Comment by asajeffrey** [comment]
Looking at the JS in FF's debugger, there are some suspicious `document.createElement("IFRAME")` in https://www.youtube.com/yts/jsbin/www-embed-player-vfl-uq_Tp/www-embed-player.js. In particular:
```js
var a=document.createElement("IFRAME");
a.style.display="none";
a.src="";
document.documentElement.appendChild(a);
```

In Servo's impl of iframes this may well create more session history entries than FF does, since the iframe is being attached with `src=""`. Since the `src` is set, https://html.spec.whatwg.org/multipage/iframe-embed-object.html#process-the-iframe-attributes says that we should navigate to `about:blank`. This causes Servo to create a new session history for the new `about:blank` page. Existing browsers may well just recycle the existing `about:blank` in this case.

We may be hitting https://github.com/whatwg/html/issues/490 here.

**Comment by asajeffrey** [comment]
@jdm: this may be more concrete evidence of web content providers (in this case youtube!) relying on a particular semantics of initial about:blank handling.

**Comment by cbrewster** [comment]
I believe this is #14720

**Comment by cbrewster** [comment]
@asajeffrey in your snippet from the embedded player, the `src` is set prior to the `iframe` being appended to the document.

The iframe will have `about:blank` in the history (as will all iframes right now, that is due to #14720). I am not certain why there are two more entries rather than just one; however, I don't think that is due to `a.src="";`.

**Comment by asajeffrey** [comment]
@cbrewster the reason why the src attribute matters is that if it wasn't set then the clause from https://html.spec.whatwg.org/multipage/iframe-embed-object.html#process-the-iframe-attributes would apply:

> Otherwise, if the element has no src attribute specified, and the user agent is processing the iframe's attributes for the "first time"
>
>    Queue a task to run the iframe load event steps.
>
>    The task source for this task is the DOM manipulation task source.

**Comment by asajeffrey** [comment]
Hmm: https://github.com/servo/servo/blob/eefd59e9332ea054908548eaba63ac12ff4bf46c/components/script/dom/htmliframeelement.rs#L259

Shouldn't that be:
```rust
self.navigate_or_reload_child_browsing_context(Some(load_data), NavigationType::Regular, true);
```

**Comment by asajeffrey** [comment]
Ah, replacing `false` by `true` in that line fixes the problem. I wonder if it breaks anything?

**Comment by asajeffrey** [comment]
IRC chat: http://logs.glob.uno/?c=mozilla%23servo&s=20+Sep+2017&e=20+Sep+2017#c755990

TL;DR: we should be OK doing: https://github.com/asajeffrey/servo/commit/5fa2dfe14ce80ef9e0607217807f8a57121a169e:
```rust
let replace = mode == ProcessingMode::FirstTime;
self.navigate_or_reload_child_browsing_context(Some(load_data), NavigationType::Regular, replace);
```

**Comment by atouchet** [comment]
I am not able to reproduce this issue in the latest nightly build using Alt + left arrow to go back.
